src/connect/mergeProps.js

```diff
@@ -12,7 +12,11 @@ export function wrapMergePropsFunc(mergeProps) {
     let mergedProps
 
     return function mergePropsProxy(stateProps, dispatchProps, ownProps) {
-      const nextMergedProps = mergeProps(stateProps, dispatchProps, ownProps)
+      let nextMergedProps = mergeProps(stateProps, dispatchProps, ownProps)
+
+      if (typeof nextMergedProps === 'function') {
+        nextMergedProps = nextMergedProps(stateProps, dispatchProps, ownProps)
+      }
```

Comment:
I don't think this is complete; it looks like it's still going to call the original factory method every time through the loop. I'd expect to see `mergeProps` copied into a variable that is overwritten with its result when the result is a function, similar to [wrapMapToPropsFunc](https://github.com/reactjs/react-redux/blob/next/src/connect/wrapMapToProps.js#L53)